cloudant.js

```diff
@@ -0,0 +1,549 @@
+          var db = cloudant.use(node.database);
+          db.bulk( { docs: doc }, function(err, body) {
+            if (err && err.status_code === 404 && attempts > 0) {
+              // status_code 404 means the database was not found
+              return cloudant.db.create(db.config.db, function() {
+                bulkDocument(cloudant, node, { docs: doc }, attempts-1, callback);
+              });
+            }
+            callback(err, body);
+          });
+        }
+    };
+    RED.nodes.registerType("cloudant out", CloudantOutNode);
+
+    function CloudantInNode(n) {
+        RED.nodes.createNode(this,n);
+
+        this.cloudantConfig = _getCloudantConfig(n);
+        this.database       = _cleanDatabaseName(n.database, this);
+        this.search         = n.search;
+        this.design         = n.design;
+        this.index          = n.index;
+        this.inputId        = "";
+
+        this.retries        = n.retries || 5; // number of connection retries
+        this.timeout        = n.timeout || 1000; // timeout between retries
+
+        var node = this;
+        var credentials = {
+            account:  node.cloudantConfig.account,
+            key:      node.cloudantConfig.username,
+            password: node.cloudantConfig.password,
+            url: node.cloudantConfig.url
+        };
+
+        node.status({fill:"red",shape:"ring",text:"disconnected"});
+
+        doConnect(credentials, 1);
+
+        function doConnect(credentials, tryNum) {
+          Cloudant(credentials, function(err, cloudant) {
+            if (err) {
+              if (tryNum <= node.retries) {
+                node.status({fill:"yellow",shape:"ring",text:"retrying..."});
+                setTimeout(function() { doConnect(credentials, tryNum + 1) }, node.timeout)
+              } else {
+                node.error(err.description, err);
+                node.status({fill:"red",shape:"dot",text:err.description});
+              }
+            } else {
+              node.status({fill:"green",shape:"dot",text:"connected"});
+              node.on("input", function(msg) {
+                if (err) { return node.error(err.description, err); }
+                handleMsg(cloudant, node, msg);
+              });
+            }
+          })
+        };
+
+
+        function handleMsg(cloudant, node, msg) {
+          var db = cloudant.use(node.database);
+          var options = (typeof msg.payload === "object") ? msg.payload : {};
+
+          if (node.search === "_id_") {
+            var id = getDocumentId(msg.payload);
+            node.inputId = id;
+
+            db.get(id, function(err, body) {
+              sendDocumentOnPayload(err, body, msg, node);
+            });
+          }
+          else if (node.search === "_idx_") {
+            console.log(options);
+            options.query = options.query || options.q || formatSearchQuery(msg.payload);
+            options.include_docs = options.include_docs || true;
```

Comment:
This one will always return true. [something falsy] || true -> true
undefined || true => true
null || true => true
false || true => true